src/main/g8/src/main/scala/worker/FrontEnd.scala

```diff
@@ -3,77 +3,80 @@ package worker
 import java.util.UUID
 import java.util.concurrent.ThreadLocalRandom
 
-import akka.actor.{Actor, ActorLogging, Cancellable, Props, Timers}
-import akka.pattern._
+import akka.actor.typed._
+import akka.actor.typed.Behavior
+import akka.actor.typed.scaladsl.Behaviors
 import akka.util.Timeout
+import worker.Master.SubmitWork
 
 import scala.concurrent.duration._
+import scala.util.Failure
+import scala.util.Success
 
 /**
- * Dummy front-end that periodically sends a workload to the master.
- */
-object FrontEnd {
-
-  def props: Props = Props(new FrontEnd)
-
-  private case object NotOk
-  private case object Tick
-  private case object Retry
-}
-
+  * Dummy front-end that periodically sends a workload to the master.
+  */
 // #front-end
-class FrontEnd extends Actor with ActorLogging with Timers {
-  import FrontEnd._
-  import context.dispatcher
-
-  val masterProxy = context.actorOf(
-    MasterSingleton.proxyProps(context.system),
-    name = "masterProxy")
+object FrontEnd {
 
-  var workCounter = 0
+  sealed trait Command
+  case object Tick extends Command
+  case object Failed extends Command
+  case object Retry extends Command
+  case object WorkAccepted extends Command
 
-  def nextWorkId(): String = UUID.randomUUID().toString
+  private def nextWorkId(): String = UUID.randomUUID().toString
 
-  override def preStart(): Unit = {
-    timers.startSingleTimer("tick", Tick, 5.seconds)
+  def apply(): Behavior[Command] = Behaviors.setup { ctx =>
+    val masterProxy = MasterSingleton.init(ctx.system)
+    idle(0, masterProxy)
   }
 
-  def receive = idle
-
-  def idle: Receive = {
-    case Tick =>
-      workCounter += 1
-      log.info("Produced work: {}", workCounter)
-      val work = Work(nextWorkId(), workCounter)
-      context.become(busy(work))
-  }
-
-  def busy(workInProgress: Work): Receive = {
-    sendWork(workInProgress)
-
-    {
-      case Master.Ack(workId) =>
-        log.info("Got ack for workId {}", workId)
-        val nextTick = ThreadLocalRandom.current.nextInt(3, 10).seconds
-        timers.startSingleTimer(s"tick", Tick, nextTick)
-        context.become(idle)
+  def idle(workCounter: Int,
+           masterProxy: ActorRef[SubmitWork]): Behavior[Command] =
+    Behaviors.withTimers { timer =>
```

Comment:
Inject timers and ctx from the apply instead of re-wrapping for every state change

Reply:
:+1: